astropy/utils/xml/src/iterparse.c

```diff
@@ -40,13 +40,56 @@
 static Py_ssize_t
 next_power_of_2(Py_ssize_t n)
 {
-    /* Calculate the next-highest power of two */
+    /* Calculate the next-higher power of two that is >= 'n' */
+
+    /* These instructions are intended for uint32_t and originally 
+       from http://www-graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+       Py_ssize_t is the same as the C datatype ssize_t and is 
+       32/64-bits on 32-bit and 64-bit systems respectively. The implementation
+       here accounts for both.
+       
+       Limitations: Since a signed size_t (ssize_t) was required for the underlying CPython implementation,
+       on 32 bit systems, it will not be possible to allocate memory sizes between
+       SSIZE_MAX and SIZE_MAX (i.e, for allocations in the range [2^31, 2^32 - 1] bytes)
+       even though such memory might be available and accessible on the (32-bit) computer. That
+       said, since the underlying CPython implementation *also* uses Py_ssize_t (i.e., ssize_t),
+       it is safe to assume that such memory allocations would probably not be usable anyway. 
+       
+       TLDR: Will work on 64-bit machines but be careful on 32 bit machines when reading in 
+       ~ 2+ GB of memory -- ManodeepS 27/03/2020
+    */
+
+    /* If negative numbers are also allowed, then we need this line: 
+    if(n <= 1) {
+        return 1;
+    } 
+    */
+    
     n--;
+
+#if 1
+    /* This section just extends the existing solution to 64 bit systems */
     n |= n >> 1;
     n |= n >> 2;
     n |= n >> 4;
     n |= n >> 8;
     n |= n >> 16;
+    if(sizeof(Py_ssize_t) > 4) {
+        n |= n >> 32; /* this works for 64-bit systems but will need to be updated if (Py)_ssize_t 
+                         ever increases beyond 64-bits */
+    }
+#else    
+    /* This is the general solution for arbitrary widths for Py_ssize_t but is much less 
+    readable and possibly over-kill */
+    const size_t numbits = sizeof(Py_ssize_t)*CHAR_BIT - 1;/* remove one for the sign-bit */
+    size_t shift = 1, totbits = 0;
+    while(totbits < numbits) {
```

Comment:
How about the following to replace the whole routine? (Also takes care of `n <= 1`)
```
size_t result = 1;
while (n > 1) {
    result <<= 1;
    n >>= 1;
}
return result;
```

Reply:
This is essentially what I used as the correct code in my [testing routine](https://gist.github.com/manodeep/110a39b43cb6472c108bc1fe13560819). The biggest difference is that the (worst-case) number of steps in this code snippet scales as the number of bits, while the other two solutions have a constant complexity of `O(log_2(nbits))`

Reply:
Ah, I now see I really didn't look carefully enough at the original code! Suggestion withdrawn!